Feature/npcbreedingstands

fiszore opened the pull request:
Allows NPCs to use any form of BreedingStand to "offload" their prey.

Uses a simple script to simulate how the player may use the stand so that all of the proper knot/distance calculations are performed, and allows for camera transitions when the player is churned.

Archival1 commented:
Reviewed this with Nael they've added some commits to refactor a few minor points.

Just to explain the camera changes. As long as the player still exists, their camera configs are treated as a stack and can be added to and removed which causes the camera to tween between configs at they are modified.

To that end the logic was altered a bit so that the source of the camera config (the cumbobulator) is responsible for creating a lambda that will remove that same config when the game object is destroyed or disabled by attaching a small monobehaviour to it that exposes that event.

Aside from the above, could I ask that you add a timer of sorts to the new simulation please? Its possible NPC's with exotic shapes or in odd spaces in maps would never go over the stimulation threshold to finish and become stuck in the stand. A timer of 30 seconds or so to force them to finish once its reached should prevent that.

Thanks again for submitting this, its a really nice feature!

fiszore commented:
Thanks for the feedback, and explanation!

I added a 35 second timer so even if the NPC reaches a stand right after catching another character, there is still a bit of a buffer after the churn is finished.

I also introduced a minimum threshold for stimulation, which should end the simulation early if a character somehow is unable to penetrate properly.

So far, I've only tested this with the standard built-in characters after increasing the stimulation reduction to prevent the threshold from being reached. I don't currently have any really exotic characters to test with, so let me know if you run into any other issues or if further changes are needed!

fiszore commented:
Finally had a chance to go back and do some additional testing, cramped spaces should no longer lock NPCs into usage even after the timer expires, and allow them to properly continue usage if multple characters have been churned. 

I also updated character stimulation to make characters hard, preventing situations where the player/NPCs could go soft during extended use of breeding stands.

I made a change to the BeCVHungry AI to prevent them from reacting to condoms containing a character they can CV. Not sure if this is a change you would want, but it prevents them from stopping use of breeding stands early and keep them from recombobulating/churning the same character repeatedly over a long period.

Archival1 commented:
I will have a chat with Nael when were both back from our Easter breaks, I think the AI changes could be simplified but I will get back to you once we've gone over it all.

Archival1 commented:
Thanks for being patient, were back!
We've gone over the recent changes you made and have some feedback.

Regarding the AI changes:
That got us thinking and we feel like it would be best if pred NPC's just ignored condoms entirely rather than trying to wrangle them into considering some but not others. Preds ignoring condoms also just feels more on brand for them!
Should be as simple as just removing the condom check entirely from the hungry action.

Regarding the boner changes:
I feel like fixing this bug would be better suited to its own pull request rather than be part of this one since the timer on the stands should prevent NPC's from getting stuck even if they go soft won't it? 
Happy to review that on its own PR if you feel like raising it.

Happy to do another round of playtesting on this branch once thats all sorted!

fiszore commented:
Hello again!

I definitely agree on preds completly ignoring condoms, what you suggested definitely fits them a lot better, and was as simple to integrate as you thought!

I also reverted the change to the boner/stimulation events as you mentioned. My idea for this change was for when an NPC had many churned characters, they would go soft before they were able to "deposit" all of them. They wouldn't get stuck thanks to the timer and usage checks, but it looked a bit... strange... to say the least.

Turns out the situation I was trying to avoid was far more situational than I thought, though I may make a PR for it at a later point.

Once again, thank you for the feedback!

Archival1 commented:
Awesome, I should be free today/tomorrow to give this a playtest!

Archival1 commented:
Ran into an issue while testing but I think its a symptom of our design rather than yours, will get back to you on that shortly but everything else is looking good!

Archival1 commented:
Just pushed an update to this that keeps the breeding stations in line with other interactables, this way its an authored choice which breeding stands a pred can use based on the group rather than allowing them to use any as this was causing npc's to leave their 'zone' to use a stand on the other side of the map.

I'm happy with where this is at so i'll do a few more tests over the next day or so and then look at merging and documenting it assuming theres nothing else you want to alter.

Thanks again for doing this Fiszore, really appreciate it! I'll have to update our existing maps to support this but we should have a patch out a week or so after this goes in to the SDK.
